### PR TITLE
MODINREACH-177 INN-Reach Record Contribution: Ignore Items and Bibs from Locations Not Associated with INN-Reach Local Agency Libraries

### PR DESCRIPTION
**Pull request by s1mpleGit**
[MODINREACH-177](https://issues.folio.org/browse/MODINREACH-177) - INN-Reach Record Contribution: Ignore Items and Bibs from Locations Not Associated with INN-Reach Local Agency Libraries

## Purpose
If a FOLIO instance's associated item records do not have an effective location that is under a FOLIO library that is associated with a FOLIO-hosted local agency (in central server configuration), then those instances and items should be automatically ignored for record contribution, regardless of other contribution criteria.

## Approach
Instances without items that are associated with a library that is assigned to a local agency code for the central server configuration associated with the contribution task are not contributed, regardless of any other record contribution criteria.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

### CODE REVIEW

File: src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java [hunk 1]
```diff
@@ -88,6 +89,22 @@ public boolean isEligibleForContribution(UUID centralServerId, Instance instance
...
+      .map(LocalAgencyDTO::getFolioLibraryIds)
+      .collect(Collectors.toList());
+
+    var result = emptyIfNull(folioLibraryIds.stream()
+      .filter(itemsEffectiveLocations::contains)
+      .collect(Collectors.toList()));
+
+    if (result.isEmpty()) {
+      return false;
+    }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java]: This check is redundant. Each item is already being validated for contribution in this method. If no eligible items found for an instance then it is not eligible as well.

File: src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java [hunk 2]
```diff
@@ -102,6 +119,19 @@ public boolean isEligibleForContribution(UUID centralServerId, Item item) {
...
+    var folioLibraryIds = centralServer.getLocalAgencies().stream()
+      .map(LocalAgencyDTO::getFolioLibraryIds)
+      .collect(Collectors.toList());
+    var result = emptyIfNull(folioLibraryIds.stream()
+      .filter(id -> id.contains(itemEffectiveLocation))
+      .collect(Collectors.toList()));
+
+    if (result.isEmpty()) {
+      return false;
+    }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java]: location is not library. See how it's mapped 
https://github.com/folio-org/mod-inn-reach/blob/b33d7e9f9966975ec8ad51f59abf57f665e85db8/src/main/java/org/folio/innreach/domain/service/impl/RecordTransformationServiceImpl.java#L221

File: src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java (hunk 1 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java]: removed

File: src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java [hunk 3]
```diff
@@ -102,6 +104,22 @@ public boolean isEligibleForContribution(UUID centralServerId, Item item) {
...
+      .map(LocalAgencyDTO::getFolioLibraryIds)
+      .collect(Collectors.toList());
+
+    var result = emptyIfNull(folioLibraryIds.stream()
+      .filter(id -> id.contains(libId))
+      .collect(Collectors.toList()));
+
+    if (result.isEmpty()) {
+      return false;
+    }
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java]:
```suggestion
    var localAgencyLibraryIds = getFolioLibraryIds(centralServerId);
    var locationLibraryMappings = folioLocationService.getLocationLibraryMappings();
    var itemLibraryId = locationLibraryMappings.get(item.getEffectiveLocationId());
    if (!localAgencyLibraryIds.contains(itemLibraryId)) {
      log.info("Item's locations is not associated with INN-Reach local agencies");
      return false;
    }
```

Comment (oliinyko) [reply src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java]: also move the logic to a private method returning boolean

Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java]: done

File: src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java (hunk 2 above)
Comment (s1mpleGit) [reply src/main/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImpl.java]: thanks for help

File: src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java [hunk 4]
```diff
@@ -50,6 +53,7 @@
   private static final UUID DO_NOT_CONTRIBUTE_CODE_ID = CRITERIA.getDoNotContributeId();
   private static final String ELIGIBLE_SOURCE = "MARC";
   private static final String INELIGIBLE_SOURCE = "FOLIO";
+  private static final UUID TEST_UUID = UUID.fromString("6802c458-b19e-476f-9187-e8ab7417ecd4");
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java]: what is TEST_UUID? Rename it to something like LOCATION_ID

File: src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java [hunk 5]
```diff
@@ -285,10 +291,14 @@ void throwWhenMultipleStatisticalCodes() {
   void testEligibleInstance_noStatisticalCodes() {
     when(contributionConfigService.getCriteria(any())).thenReturn(CRITERIA);
     when(holdingsService.find(any())).thenReturn(Optional.empty());
+    when(folioLocationService.getLocationLibraryMappings()).thenReturn(Map.of(TEST_UUID, TEST_UUID));
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java]: return different values for location and library ids

File: src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java [hunk 6]
```diff
@@ -303,16 +313,38 @@ void testEligibleInstance_statisticalCodeAllowed() {
...
     instance.setSource(ELIGIBLE_SOURCE);
-    instance.setItems(List.of(new Item().statisticalCodeIds(statisticalCodes)));
+    instance.setItems(List.of(new Item().statisticalCodeIds(statisticalCodes).effectiveLocationId(TEST_UUID)));
 
     when(contributionConfigService.getCriteria(any())).thenReturn(CRITERIA);
     when(holdingsService.find(any())).thenReturn(Optional.empty());
+    when(folioLocationService.getLocationLibraryMappings()).thenReturn(Map.of(TEST_UUID, TEST_UUID));
+    when(centralServerService.getCentralServer(any()))
+      .thenReturn(new CentralServerDTO().id(UUID.randomUUID()).localAgencies(List.of(
+        new LocalAgencyDTO().id(UUID.randomUUID()).folioLibraryIds(List.of(TEST_UUID)))));
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java]: try to use CentralServerFixture.createCentralServer() here and for other tests

File: src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java (hunk 4 above)
Comment (s1mpleGit) [reply src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java]: done

File: src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java (hunk 5 above)
Comment (s1mpleGit) [reply src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java]: done

File: src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java (hunk 6 above)
Comment (s1mpleGit) [reply src/test/java/org/folio/innreach/domain/service/impl/ContributionValidationServiceImplTest.java]: added

**Review by oliinyko — APPROVED** [review]

**Pull request merged**